packages/json-rpc/ws-provider/src/ws-provider.ts

```diff
@@ -1,56 +1,185 @@
+  event: any
+}
+export type WsClose = {
+  type: WsEvent.CLOSE
+  event: any
+}
+export type StatusChange = WsConnecting | WsConnected | WsError | WsClose
+export type WsJsonRpcProvider = JsonRpcProvider & {
+  switch: (uri?: string, protocol?: string[]) => void
+  getStatus: () => StatusChange
+}
+
+export interface GetWsProviderInput {
+  (
+    uri: string,
+    protocols?: string | string[],
+    onStatusChanged?: (status: StatusChange) => void,
+  ): WsJsonRpcProvider
+  (
+    uri: string,
+    onStatusChanged?: (status: StatusChange) => void,
+  ): WsJsonRpcProvider
+  (
+    endpoints: Array<string | { uri: string; protocol: string[] }>,
+    onStatusChanged?: (status: StatusChange) => void,
+  ): WsJsonRpcProvider
+}
+
+export const getInternalWsProvider = (
+  WebsocketClass: typeof WebSocket,
+): GetWsProviderInput => {
+  return (...args): WsJsonRpcProvider => {
+    let endpoints: Array<[string, string | string[]] | [string]> = []
+    let onStatusChanged: (status: StatusChange) => void = () => {}
 
+    if (typeof args[1] === "function") onStatusChanged = args[1]
+    if (Array.isArray(args[0]))
+      endpoints = args[0].map((x) =>
+        typeof x === "string" ? [x] : [x.uri, x.protocol],
+      )
+    else {
+      endpoints = [[args[0]]]
+      if (args[1]) {
+        if (args[1] !== onStatusChanged) endpoints[0][1] = args[1] as any
+        else if (args[2]) onStatusChanged = args[2] as any
+      }
+    }
+    let idx = 0
+    let status: StatusChange
+    let switchTo: [string] | [string, string | string[]] | null = null
+    let disconnect: (withHalt?: boolean) => void = () => {}
+
+    const result = getSyncProvider(async () => {
+      const [uri, protocols] = switchTo || endpoints[idx++ % endpoints.length]
+      switchTo = null
+      const socket = new WebsocketClass(uri, protocols)
+      onStatusChanged(
+        (status = {
+          type: WsEvent.CONNECTING,
+          uri,
+          protocols,
+        }),
+      )
       await new Promise<void>((resolve, reject) => {
         const onOpen = () => {
+          cleanup()
           resolve()
-          socket.removeEventListener("error", onError)
         }
-        socket.addEventListener("open", onOpen, { once: true })
 
         const onError = (e: Event) => {
+          cleanup()
           console.error(
             `Unable to connect to ${uri}${
               protocols ? ", protocols: " + protocols : ""
             }`,
           )
-          reject(e)
+          onStatusChanged(
+            (status = {
+              type: e ? WsEvent.ERROR : WsEvent.CLOSE,
+              event: e,
+            }),
+          )
+          setTimeout(reject, e ? 300 : 0, e)
+        }
+        const timeoutToken = setTimeout(() => {
+          const event = { type: "timeout" }
+          onStatusChanged(
+            (status = {
+              type: WsEvent.ERROR,
+              event,
+            }),
+          )
+          reject(event)
+        }, 1_000)
```

Comment:
(related to previous question) a timeout of 1 second, maybe that's a bit too aggressive?

Reply:
for establishing the initial connection? I don't think so... I mean, it normally is pretty much immediate. Even if the internet connection is bad, it should't take more than a few hundred of milliseconds. Let's keep in mind that this is the timeout for establishing the initial connection, not for receiving the data.

Reply:
As we discussed offline, I ended up increasing it to 3.5s. Thanks!